### PR TITLE
Product Frontend

### PR DESCRIPTION
**Pull request by aShahriarSazim**
Created Product CRUD frontend

### CODE REVIEW

File: client/src/store/slices/ProductSlice/ProductSlice.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import { createSlice, PayloadAction } from '@reduxjs/toolkit';
+import  ProductInterface  from './Interfaces/Product/productInterface';
+
+const initialState : ProductInterface[]  = [];
```
Comment (rChowdhurySazim) [inline client/src/store/slices/ProductSlice/ProductSlice.ts:4]: You can make it an object with { allProduct : [], userProducts: [], etc.. } when you use createAsyncThunk from RTK

File: client/src/App.tsx [hunk 2]
```diff
@@ -5,12 +5,14 @@ import store from "./store/index";
...
 import {BrowserRouter} from "react-router-dom";
 import CheckAuth from "./components/CheckAuth";
+import FetchAllProducts from "./components/FetchAllProducts";
 
 const App: FC = () => {
     return (
         <Provider store={store}>
             <BrowserRouter>
                 <CheckAuth />
+                <FetchAllProducts />
```
Comment (rChowdhurySazim) [inline client/src/App.tsx:15]: Should not do this. Fetch products in components where its necessary also should not use a whole component just for the fetching in useEffect.

File: client/src/components/FetchAllProducts.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import {FC, useEffect} from "react";
+import {useAppDispatch, useAppSelector} from "../store/hooks";
+import axios from "../axios";
+import {setAllProducts} from "../store/slices/ProductSlice";
+
+const FetchAllProducts: FC = () => {
```
Comment (rChowdhurySazim) [inline client/src/components/FetchAllProducts.tsx:6]: - You need to use createAsyncThunk from RTK do all kind of api calls and manage the state in that slice
- Should not fetch all products with just one endpoint and do all operations on that data
- Need other endpoints like user/5/products or something to return the products owned by the user
- FetchAll should return the list of the products that are not owned by the user
- All kind of api calls should be in their respective files. There should not be a separate component just to use useEffect and state.

File: client/src/pages/Products/CreateProduct/CreateProduct.tsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        }else{
+            let newProduct = {
+                title: data.title,
+                description: data.description,
+                price: parseFloat(String(data.price)),
+                categories: inputCategories,
+                rent: parseFloat(String(data.rent)),
+                rentPaymentPeriod: data.rentPaymentPeriod,
+            };
+            // console.log(newProduct);
```
Comment (rChowdhurySazim) [inline client/src/pages/Products/CreateProduct/CreateProduct.tsx:47]: remove comment

File: client/src/pages/Products/CreateProduct/CreateProduct.tsx [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            const response = await axios.post("/products/create", newProduct);
```
Comment (rChowdhurySazim) [inline client/src/pages/Products/CreateProduct/CreateProduct.tsx:48]: - You need to use createAsyncThunk from RTK do all kind of api calls and manage the state in that slice

File: client/src/pages/Products/EditProduct/EditProduct.tsx [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        }else{
+            let productToBeUpdated = {
+                title: data.title,
+                description: data.description,
+                price: parseFloat(String(data.price)),
+                categories: inputCategories,
+                rent: parseFloat(String(data.rent)),
+                rentPaymentPeriod: data.rentPaymentPeriod,
+            };
+            const response = await axios.post(`/products/update/${id}`, productToBeUpdated);
```
Comment (rChowdhurySazim) [inline client/src/pages/Products/EditProduct/EditProduct.tsx:50]: - You need to use createAsyncThunk from RTK do all kind of api calls and manage the state in that slice

File: client/src/pages/Products/MyProducts/MyProducts.tsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import axios from "../../../axios";
+import Product from "../components/Product";
+import {useAppSelector} from "../../../store/hooks";
+import {Link, useNavigate} from "react-router-dom";
+
+const MyProducts: FC = () => {
+    const auth = useAppSelector(state => state.auth);
+    const myProducts = useAppSelector(state => state.products).filter(
+        product => product.user.email === auth.email
+    );
```
Comment (rChowdhurySazim) [inline client/src/pages/Products/MyProducts/MyProducts.tsx:13]: - Need other endpoints like user/5/products or something to return the products owned by the user

**Review by rChowdhurySazim — CHANGES_REQUESTED** [review]
Please go through the comments and address them. One comment issue might be in multiple places, so address them everywhere.

File: client/src/pages/Products/components/Product.tsx [hunk 8]
```diff
@@ -0,0 +1,80 @@
+import {DeleteIcon} from "@chakra-ui/icons";
```
[the remark this answers is not in the document]
Comment (aShahriarSazim) [reply client/src/pages/Products/components/Product.tsx:1]: But view product page will have a few more functionalities (buy and rent) later

File: client/src/pages/Products/CreateProduct/CreateProduct.tsx [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+            const createdProduct = response.data;
+            dispatch(addProduct(createdProduct));
+            navigateTo('/products/my');
+        }
+    };
+    const categoryInputUpdated = (e: any) => {
+        let temp: number[] = [];
+        e.forEach((category: categoryOptions) => {
+            temp.push(category.value);
```
[the remark this answers is not in the document]
Comment (aShahriarSazim) [reply client/src/pages/Products/CreateProduct/CreateProduct.tsx:52]: done

File: client/src/pages/Products/CreateProduct/CreateProduct.tsx [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+        });
+        setInputCategories(temp);
+    };
+    useEffect(() => {
+    }, []);
```
[the remark this answers is not in the document]
Comment (aShahriarSazim) [reply client/src/pages/Products/CreateProduct/CreateProduct.tsx:57]: removed unnecessary codes

File: client/src/pages/Products/Products.tsx [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+import {Box, Text} from "@chakra-ui/react";
+import Product from "./components/Product";
+import {useAppSelector} from "../../store/hooks";
+
+const Products: FC = () => {
+
+    const allProducts = useAppSelector(state => state.products);
+
+    useEffect(() => {
+    }, []);
```
[the remark this answers is not in the document]
Comment (aShahriarSazim) [reply client/src/pages/Products/Products.tsx:15]: removed unnecessary codes

File: client/src/pages/Products/ViewProduct/ViewProduct.tsx [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+import {useAppDispatch, useAppSelector} from "../../../store/hooks";
+import axios from "../../../axios";
+import {removeProduct} from "../../../store/slices/ProductSlice";
+
+const ViewProduct: FC = () => {
+    const {id} = useParams();
+    const dispatch = useAppDispatch();
+    const auth = useAppSelector(state => state.auth);
+    // @ts-ignore
+    let product = useAppSelector(state => state.products.find(product => (product.id == parseInt(id))));
```
[the remark this answers is not in the document]
Comment (aShahriarSazim) [reply client/src/pages/Products/ViewProduct/ViewProduct.tsx]: I haven't made those api endpoints yet, so I have to solve them in the next branch (should I make a new branch off of this product-FE branch and resolve them there and then merge?

File: client/src/pages/Products/components/Product.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (aShahriarSazim) [reply client/src/pages/Products/components/Product.tsx]: changed the name to ProductCard

File: client/src/App.tsx (hunk 2 above)
Comment (aShahriarSazim) [reply client/src/App.tsx]: removed this

File: client/src/pages/Products/ViewProduct/ViewProduct.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (aShahriarSazim) [reply client/src/pages/Products/ViewProduct/ViewProduct.tsx]: Resolved

File: client/src/components/FetchAllProducts.tsx (hunk 3 above)
Comment (aShahriarSazim) [reply client/src/components/FetchAllProducts.tsx]: Done

File: client/src/pages/Products/CreateProduct/CreateProduct.tsx (hunk 4 above)
Comment (aShahriarSazim) [reply client/src/pages/Products/CreateProduct/CreateProduct.tsx]: Done

File: client/src/pages/Products/CreateProduct/CreateProduct.tsx (hunk 5 above)
Comment (aShahriarSazim) [reply client/src/pages/Products/CreateProduct/CreateProduct.tsx:46]: Done

File: client/src/pages/Products/EditProduct/EditProduct.tsx (hunk 6 above)
Comment (aShahriarSazim) [reply client/src/pages/Products/EditProduct/EditProduct.tsx:50]: Done

File: client/src/pages/Products/MyProducts/MyProducts.tsx (hunk 7 above)
Comment (aShahriarSazim) [reply client/src/pages/Products/MyProducts/MyProducts.tsx]: Done

**Review by rChowdhurySazim — APPROVED** [review]

**Pull request merged**